Add support for Quantization-Aware Low-Rank Adaptation (QALoRA)

gapsong opened the pull request:
This pull request introduces QALoRA (Quantization-Aware Low-Rank Adaptation), a new fine-tuning technique for quantized large language models, along with its implementation in the PEFT library. The changes include updates to documentation, configuration, and core logic to support QALoRA's memory-efficient and performance-preserving features.

### Documentation Updates
* [`examples/qalora_finetuning/README.md`](diffhunk://#diff-6e8694feed4e98d9caf8f6a187aeb945e989bf6e00392be4b7674bb0b6be96d7R1-R178): Added detailed documentation for QALoRA, including its introduction, implementation details, usage examples, command-line instructions, and comparison with other techniques like LoRA and DoRA.

### Configuration Enhancements
* [`src/peft/tuners/lora/config.py`](diffhunk://#diff-13380145401d203d5935c5189dd09879f990b81aa63e8e3aaff8ce9110333f0eR497-R519): Introduced two new configuration parameters: `use_qalora` to enable QALoRA and `qalora_group_size` to control the pooling group size for memory-performance tradeoffs.

### Core Logic for QALoRA
* [`src/peft/tuners/lora/gptq.py`](diffhunk://#diff-c65b84680649207b662504c2e0bcf65d8478ae27ed4ff6d7d58f06b615b025a2R22): Updated the GPTQ LoRA implementation to support QALoRA, including logic for resolving QALoRA variants and passing group size parameters. [[1]](diffhunk://#diff-c65b84680649207b662504c2e0bcf65d8478ae27ed4ff6d7d58f06b615b025a2R22) [[2]](diffhunk://#diff-c65b84680649207b662504c2e0bcf65d8478ae27ed4ff6d7d58f06b615b025a2R36) [[3]](diffhunk://#diff-c65b84680649207b662504c2e0bcf65d8478ae27ed4ff6d7d58f06b615b025a2R58-R74)
* [`src/peft/tuners/lora/layer.py`](diffhunk://#diff-24a141c266b7b714ae8fcc470f31bc283f7b0f5a671bbf6d5f092741fc374104R187-R208): Enhanced the layer update logic to initialize QALoRA-specific parameters and handle adapter-specific configurations. [[1]](diffhunk://#diff-24a141c266b7b714ae8fcc470f31bc283f7b0f5a671bbf6d5f092741fc374104R187-R208) [[2]](diffhunk://#diff-24a141c266b7b714ae8fcc470f31bc283f7b0f5a671bbf6d5f092741fc374104L242-R253)
* [`src/peft/tuners/lora/model.py`](diffhunk://#diff-4331e1b00e557c2af6e682bc221d1517f554d7ca24b9149520237b431d8d797fR203-R204): Incorporated QALoRA-specific parameters into the model creation and replacement process.

### QALoRA Variant Implementation
* [`src/peft/tuners/lora/variants.py`](diffhunk://#diff-a14f8c2e9d3ff608a836f59d47401dc0e66c19a8e7ea5c25434ab857da6c0eccR133-R321): Added the `QALoraLinearVariant` class, implementing QALoRA-specific logic for initialization, delta weight computation, merging, unmerging, and forward propagation. This includes pooling input features and scaling them for efficient adaptation.

BenjaminBossan commented on src/peft/tuners/lora/gptq.py:
```diff
@@ -53,9 +55,23 @@ def __init__(
         )
 
+    def resolve_lora_variant(self, *, use_dora: bool, use_qalora: bool, **kwargs) -> Optional[LoraVariant]:
+        if use_qalora:
+            from .variants import QALoraLinearVariant
+
+            return QALoraLinearVariant()
+        if not use_dora:
+            return None
+
+        from .variants import DoraLinearVariant
+
+        return DoraLinearVariant()
```
Let's change the check a bit for completeness to basically:

```python
if use_dora and use_qalora:
    NotImplementedError
elif use_dora:
    variant = ...
elif use_qalora:
    variant = ...
else:
    variant = None
return variant
```

BenjaminBossan commented on src/peft/tuners/lora/gptq.py:
```diff
@@ -64,29 +80,33 @@ def forward(self, x: torch.Tensor):
             return result
 
         lora_A_keys = self.lora_A.keys()
+        torch_result_dtype = result.dtype
```
This is not needed, right?

BenjaminBossan commented on src/peft/tuners/lora/gptq.py:
```diff
@@ -64,29 +80,33 @@ def forward(self, x: torch.Tensor):
-
-            if requires_conversion:
-                output = output.to(expected_dtype)
+            # requires_conversion = not torch.is_autocast_enabled()
```
remove the comment?

BenjaminBossan commented on src/peft/tuners/lora/layer.py:
```diff
@@ -184,17 +184,28 @@ def update_layer(
-        # collect the kwargs
-        kwargs = locals().copy()
-        del kwargs["self"]
+        params_for_variant_init = {
```
Why was this change necessary?

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+        """
+        # Store the qalora_group_size as a module attribute
+        # It's expected that qalora_group_size is passed in kwargs from LoraConfig
+        qalora_group_size = kwargs.get("qalora_group_size", 16)  # Default to 16 if not specified in kwargs
```
We should ensure the argument is always there, thus `kwargs.get` should not be necessary.

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+        # It's expected that qalora_group_size is passed in kwargs from LoraConfig
+        qalora_group_size = kwargs.get("qalora_group_size", 16)  # Default to 16 if not specified in kwargs
+        if not hasattr(module, "qalora_group_size"):
+            module.qalora_group_size = {}
```
We need to add: `module.other_param_names = module.other_param_names[:] + ("qalora_group_size",)`

This corresponds to this DoRA code:

https://github.com/huggingface/peft/blob/e67052b18cb3c58370e0339ecf76a5faf447f2ff/src/peft/tuners/lora/variants.py#L33

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+
+        # Store original dimensions for use in get_delta_weight
+        if not hasattr(module, "orig_in_features"):
+            module.orig_in_features = {}
```
Same comment as above regarding `other_param_names`

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+
+        # Create and store pooling factor for scaling
+        if not hasattr(module, "qalora_scaling_factor"):
+            module.qalora_scaling_factor = {}
```
Same comment as above regarding `other_param_names`. But do we even need `qalora_scaling_factor`, as it can be calculated on the fly based on `module.in_features` and `qalora_group_size` anyway?

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+            module.qalora_scaling_factor[adapter_name] = module.in_features / qalora_group_size
+        else:
+            # No special scaling if dimensions don't align
+            module.qalora_scaling_factor[adapter_name] = 1.0
```
Would this not lead to very different results than `qalora_scaling_factor = module.in_features / qalora_group_size`?

I wonder if perhaps it makes more sense to raise an error here and require that `module.in_features % qalora_group_size == 0`? I think this would simplify the code and when `module.in_features % qalora_group_size != 0` we just get standard LoRA?

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+            torch.Tensor: The calculated delta weight.
+        """
+        if (
+            not hasattr(module, "qalora_group_size")
```
Is it even possible to hit this condition? Same question about `not hasattr(module, "qalora_scaling_factor")`.

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,186 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+            module.qalora_scaling_factor[adapter_name] = 1.0
+
+    @staticmethod
+    def get_delta_weight(module: Linear, active_adapter: str) -> torch.Tensor:
```
Just wondering, as merging is not implemented, do we even need this method? I guess we could leave it here in case merging is added later, but at least a comment should be added in that case.

BenjaminBossan commented on examples/qalora_finetuning/qalora_gptq_finetuning.py:
```diff
@@ -0,0 +1,422 @@
+from peft import LoraConfig, get_peft_model
+
+
+def check_adapter_gradients(model) -> None:
```
I think this type of check could be moved to a unit test instead.

BenjaminBossan commented on examples/qalora_finetuning/qalora_gptq_finetuning.py:
```diff
@@ -0,0 +1,422 @@
+    for name, param in model.named_parameters():
+        param_size = param.numel() * param.element_size() / 1024**2  # in MB
+        grad_size = param_size if param.requires_grad else 0
+        opt_size = 0
```
If the user uses neither Adam nor SGD, this would be shown as 0, which would be confusing. Instead, you could check if the value is 0 and then replace it, for instance, with `"unknown"`.

BenjaminBossan commented on examples/qalora_finetuning/qalora_gptq_finetuning.py:
```diff
@@ -0,0 +1,422 @@
+
+    # Get PEFT model with adapters
+    model = get_peft_model(model, lora_config)
+    model.save_pretrained(output_dir)
```
Why save the checkpoint here?

BenjaminBossan commented on examples/qalora_finetuning/qalora_gptq_finetuning.py:
```diff
@@ -0,0 +1,422 @@
+        f"\nTotal gradient memory (training): {total_grad_mem:.2f} MB"
+        f"\nTotal optimizer state memory (training): {total_opt_mem:.2f} MB"
+        f"\nTotal memory for training: {total_param_mem + total_grad_mem + total_opt_mem:.2f} MB"
+        f"\nTotal memory for inference: {total_param_mem:.2f} MB"
```
I think there is an error somewhere in the calculation. When running this with `TheBloke/Llama-2-7b-GPTQ`, I get:

```
Parameter dtype is per-column above. 
Total parameter memory: 576.76 MB
Total gradient memory (training): 76.25 MB
Total optimizer state memory (training): 152.50 MB
Total memory for training: 805.51 MB
Total memory for inference: 576.76 MB
```

These values are too low, both from a theoretical viewpoint and from what I measure in practice (of course, in practice we need memory for hidden states, which are difficult to estimate).

When I use [my own script](https://github.com/BenjaminBossan/presentations/blob/main/2024-07-11-europython/utils.py) for checking, with int4 quantized `meta-llama/Llama-2-7b-hf`, Adam, and LoRA rank 16, what I get is:

```json
{
  "number of parameters": {
    "Embedding.weight": 131072000,
    "Linear.weight": 6476005376,
    "lora": 39976960,
    "LlamaRMSNorm.weight": 266240
  },
  "number of bytes": {
    "Embedding.weight": 262144000,
    "Linear.weight": 3238002688,
    "lora": 159907840,
    "LlamaRMSNorm.weight": 532480
  },
  "number of bytes (readable)": {
    "Embedding.weight": "250.0 MB",
    "Linear.weight": "3.02 GB",
    "lora": "152.5 MB",
    "LlamaRMSNorm.weight": "520.0 KB"
  },
  "total number of parameters w/o LoRA": 6607343616,
  "total number of parameters w/  LoRA": 6647320576,
  "total size w/o LoRA": 3500679168,
  "total size w/  LoRA": 3660587008,
  "total size w/o LoRA (readable)": "3.26 GB",
  "total size w/  LoRA (readable)": "3.41 GB",
  "memory required for full fine-tuning": "13.04 GB*",
  "memory required for LoRA fine-tuning": "3.71 GB"
}
```
This does not take into account QALoRA, but I don't think the difference can be that big.

BenjaminBossan commented on examples/qalora_finetuning/qalora_gptq_finetuning.py:
```diff
@@ -0,0 +1,422 @@
+            elif optimizer_name.lower() == "sgd":
+                opt_size = param_size  # SGD with momentum keeps 1 extra state
+
+        print(
```
Again, I think printing this for each layer is a bit of information overkill.

BenjaminBossan commented on examples/qalora_finetuning/README.md:
````diff
@@ -0,0 +1,178 @@
+
+Run the finetuning script with a GPTQ quantized model:
+```bash
+python examples/qalora_finetuning/qalora_gptq_finetuning.py \
````
When I run this locally, I get a train loss of 0.0, can you replicate?

BenjaminBossan commented on src/peft/tuners/lora/gptq.py:
```diff
@@ -53,9 +56,25 @@ def __init__(
 
+    def resolve_lora_variant(self, *, use_dora: bool, use_qalora: bool, **kwargs) -> Optional[LoraVariant]:
+        if use_dora and use_qalora:
+            NotImplementedError
```
```python
            raise NotImplementedError(...)
```

BenjaminBossan commented on src/peft/tuners/lora/gptq.py:
```diff
@@ -53,9 +56,25 @@ def __init__(
+            from .variants import DoraLinearVariant
+            variant = DoraLinearVariant()
+        elif use_qalora:
+            if self.in_features % kwargs["qalora_group_size"] == 0:
```
Let's not perform the check here. Instead, let's move it inside `QALoraLinearVariant.init`.

BenjaminBossan replied:
This check can now be removed, right?

BenjaminBossan reviewed (changes requested):
Thanks a lot for all the updates. I still found a couple of issues, please check my comments. I also checked and ran the script, where I also found some areas for improvement.

As for testing, I think it would be enough to have something very similar to [this one](https://github.com/huggingface/peft/blob/759bb70acea8486b3a493792a5b6ba80b5217fb4/tests/test_gpu_examples.py#L1748) with QALoRA enabled.

Also, before pushing, don't forget to run `make style` to make the linter happy.

gapsong commented:
More tests are still needed for comprehensive coverage and some of the existing comments are a work in progress and will be refined.

BenjaminBossan commented:
> More tests are still needed for comprehensive coverage and some of the existing comments are a work in progress and will be refined.

Okay, ich warte dann, bis diese fertig sind bevor ich erneut reviewe ;)

gapsong commented:
@BenjaminBossan du kannst wieder reinschauen.

Ich habe die Initialisierung nochmal überarbeitet. Jetzt wird lora_A direkt beim Init durch eine neue Instanz mit der gepoolten Eingabedimension ersetzt. Vorher hatten wir lora_A noch in der alten Shape und haben das Pooling erst im Forward angewendet, so wie es auch ursprünglich im Paper gemacht wurde.

Mit der neuen Variante übernimmt lora_A das Pooling bereits beim Init, wodurch die Dimensionen und damit die Anzahl der trainierbaren Parameter bei QALoRA tatsächlich kleiner werden als bei QLoRA. Vorher waren die "trainable params" bei QLoRA und QALoRA identisch, jetzt ist QALoRA sparsamer.

Eine interessante Beobachtung: Der Speicherverbrauch ist trotzdem sehr ähnlich zu QLoRA, obwohl weniger trainierbare Parameter vorhanden sind. Das scheint aber durch die zusätzlichen Pooling-Tensoren im Forward zustande zu kommen. Vielleicht kannst du mir nochmal bestätigen, ob das so korrekt und zu erwarten ist.

BenjaminBossan commented on examples/qalora_finetuning/README.md:
```diff
@@ -0,0 +1,178 @@
+
+## Introduction
+[QALoRA](https://huggingface.co/papers/2309.14717) is a quantization-aware version of Low-Rank Adaptation that enables efficient fine-tuning of quantized large language models. QALoRA uses input feature pooling and a specialized grouping technique to work with quantized weights, significantly reducing memory requirements while preserving performance. QALoRA enables fine-tuning of models that woul ...
+
```
Still missing?

BenjaminBossan commented on tests/test_gpu_examples.py:
```diff
@@ -1168,201 +1168,32 @@ def test_initialize_dora_with_bnb_on_cpu(self, kbit):
         assert not weights_not_cpu
 
-    @pytest.mark.single_gpu_tests
-    def test_causal_lm_training_vera(self):
```
Hmm, it looks like a bunch of unrelated tests where moved around. I don't think that should be necessary and it also makes reviewing a lot harder. Could you please ensure that the existing tests remain in place and only new tests for GPTQ-QALoRA are added?

BenjaminBossan commented:
@gapsong Please ping me once the PR is ready for the next review.

BenjaminBossan commented:
Thanks for the updates. I tried the example script again for 5 steps, using `TheBloke/Llama-2-7B-GPTQ`, and found that the `train_loss` is 0.0. Can you confirm that? It doesn't matter if `use_qalora` is enabled or not.

> Could you convince me that `if self.in_features % kwargs["qalora_group_size"] == 0:` belongs into the `init()`?

For, mostly for these reasons:

1. These types of checks are typically done during init methods in PEFT, so it is consistent.
2. It is a QALoRA related piece of code, thus it makes sense to put in the QALoRA specific code.
3. `resolve_lora_variant` should focus on what the name says, deciding which LoRA variant is used.

gapsong commented:
@BenjaminBossan I’m done and I found the “bug”:
The qalora_group_size was set to 8, which was too small. I’ve changed the default value to 32.

I also adjusted the resolve_lora_variant function:
Training with "lora" is no longer the default if the dimensions do not fit—specifically, when module.in_features % kwargs["qalora_group_size"] != 0. Instead, an error is now thrown if the parameters do not match.

I hope this time works out :)

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,98 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+        """
+        if "qalora_group_size" not in kwargs:
+            raise ValueError(
+                "QALoraLinearVariant.init expects 'qalora_group_size' to be provided in kwargs."
```
```suggestion
                "`use_qalora=True` requires 'qalora_group_size' to be provided in kwargs."
```

BenjaminBossan commented on src/peft/tuners/lora/variants.py:
```diff
@@ -130,6 +130,98 @@ def forward(module: Linear, active_adapter: str, x: torch.Tensor, result: torch.
+
+        if module.in_features is not None and module.in_features % kwargs["qalora_group_size"] != 0:
+            raise ValueError(
+                f"QALoraLinearVariant.init expects module.in_features ({module.in_features}) to be divisible by 'qalora_group_size' ({kwargs['qalora_group_size']})"
```
```suggestion
                f"`use_qalora=True` requires `module.in_features` ({module.in_features}) to be divisible by 'qalora_group_size' ({kwargs['qalora_group_size']})"
```

gapsong commented:
@BenjaminBossan something seems off with the GPU memory, when using QA_Lora I am investigating the problem at the moment.

gapsong commented:
@BenjaminBossan GPU problem was resolved. I removed `torch.nn.functional.avg_pool1d` since the peak memory needed for the unsqueezed layers were to high

BenjaminBossan commented on tests/test_custom_models.py:
```diff
@@ -2820,6 +2820,51 @@ def test_requires_grad_lora_different_targets(self):
             "base_model.model.lin1.lora_B.adapter1.weight",
         )
 
+    def test_requires_grad_qalora_same_targets(self):
```
Hmm, for some reason I haven't really noticed this test. IMO it is not necessary and can be removed. This is because QALoRA does not change the handling of multiple adapters, so there is not really a reason to believe there could be anything wrong there. We also don't check other quantization methods here. Unless you had a specific reason to add this test, I'd suggest to just remove it.

BenjaminBossan reviewed (commented):
Thanks for the latest update. I just have a small comment, otherwise the PR LGTM.

I also ran the example locally with and without QALoRA and the difference in memory usage was negligible, which I think is what we hoped to see.

HuggingFaceDocBuilderDev commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/peft/pr_2571). All of your documentation changes will be reflected on that endpoint. The docs are available until 30 days after the last update.

BenjaminBossan reviewed (changes requested):
I did a final pass and discovered two small things that need fixing, otherwise LGTM. Thanks for your patience.

gapsong commented:
@BenjaminBossan shall we merge it then?

gapsong commented:
> > shall we merge it then?
> 
> Did you see my [last comment](https://github.com/huggingface/peft/pull/2571#pullrequestreview-2957301813)?

I have missed it. It should be ready now

thistleknot commented:
thank you, now training with qa-lora =D
